File: coding-labs/merge-k-sorted-lists/java/src/main/java/MergeKSortedLists.java
import java.util.Comparator;
import java.util.List;
import java.util.PriorityQueue;

public final class MergeKSortedLists {
  private MergeKSortedLists() {}

  public static final class ListNode {
    public final int value;
    public ListNode next;

    public ListNode(int value) {
      this.value = value;
    }
  }

  private record Entry(int value, long sequence, ListNode node) {}

  /** Returns a newly allocated sorted list without rewiring caller-owned nodes. */
  public static ListNode merge(List<ListNode> lists) {
    if (lists == null) throw new IllegalArgumentException("lists are required");

    PriorityQueue<Entry> heap = new PriorityQueue<>(
        Comparator.comparingInt(Entry::value).thenComparingLong(Entry::sequence));
    long sequence = 0;
    for (ListNode node : lists) {
      if (node != null) heap.add(new Entry(node.value, sequence++, node));
    }

    ListNode sentinel = new ListNode(0);
    ListNode tail = sentinel;
    while (!heap.isEmpty()) {
      Entry entry = heap.remove();
      tail.next = new ListNode(entry.node.value);
      tail = tail.next;
      if (entry.node.next != null) {
        heap.add(new Entry(entry.node.next.value, sequence++, entry.node.next));
      }
    }
    return sentinel.next;
  }
}

File: coding-labs/merge-k-sorted-lists/python/src/merge_k_sorted_lists.py
from __future__ import annotations

from dataclasses import dataclass
from heapq import heappop, heappush
from itertools import count
from typing import Sequence


@dataclass
class ListNode:
    value: int
    next: ListNode | None = None


def merge_k_lists(lists: Sequence[ListNode | None]) -> ListNode | None:
    """Return a new sorted list without mutating caller-owned nodes."""
    if lists is None:
        raise ValueError("lists are required")

    sequence = count()
    heap: list[tuple[int, int, ListNode]] = []
    for node in lists:
        if node is not None:
            heappush(heap, (node.value, next(sequence), node))

    sentinel = ListNode(0)
    tail = sentinel
    while heap:
        _, _, node = heappop(heap)
        tail.next = ListNode(node.value)
        tail = tail.next
        if node.next is not None:
            heappush(heap, (node.next.value, next(sequence), node.next))
    return sentinel.next
